File: library/src/main/java/com/qingmei2/rximagepicker/delegate/ProxyProviders.java
package com.qingmei2.rximagepicker.delegate;

import com.qingmei2.rximagepicker.core.IRxImagePickerProcessor;
import com.qingmei2.rximagepicker.core.RxImagePicker2;
import com.qingmei2.rximagepicker.di.DaggerRxImagePickerComponent;
import com.qingmei2.rximagepicker.di.RxImagePickerModule;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

import io.reactivex.Observable;

/**
 * Created by qingmei2 on 2018/1/13.
 */
public final class ProxyProviders implements InvocationHandler {

    private final IRxImagePickerProcessor rxImagePickerProcessor;
    private final ProxyTranslator proxyTranslator;

    public ProxyProviders(RxImagePicker2.Builder builder,
                          Class<?> providersClass) {
        rxImagePickerProcessor = DaggerRxImagePickerComponent.builder()
                .rxImagePickerModule(new RxImagePickerModule(builder))
                .build()
                .rxImagePickerProcessor();

        proxyTranslator = new ProxyTranslator();
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        Class<?> methodType = method.getReturnType();
        if (methodType != Observable.class) {
            throw new IllegalArgumentException("the RxImagePicker2's return value's type only supports the Observable now.");
        }
        return rxImagePickerProcessor.process(proxyTranslator.processMethod(method, args));
    }

}

File: rximagepicker/src/main/java/com/qingmei2/rximagepicker/core/ProxyProviders.kt
package com.qingmei2.rximagepicker.core

import com.qingmei2.rximagepicker.scheduler.RxImagePickerSchedulers
import io.reactivex.*
import io.reactivex.functions.Supplier
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

class ProxyProviders : InvocationHandler {

    private val rxImagePickerProcessor = ConfigProcessor(
            RxImagePickerSchedulers()
    )
    private val proxyTranslator = ProxyTranslator()

    override fun invoke(proxy: Any, method: Method, args: Array<Any>?): Any {

        return Observable.defer(Supplier<ObservableSource<*>> {
            val configProvider = proxyTranslator.processMethod(method, args)

            ImagePickerController(configProvider).display()

            val observable = rxImagePickerProcessor.process(configProvider)

            val methodType = method.returnType

            if (methodType == Observable::class.java)
                return@Supplier Observable.just(observable)

            if (methodType == Single::class.java)
                return@Supplier Observable.just<Single<*>>(Single.fromObservable(observable))

            if (methodType == Maybe::class.java)
                return@Supplier Observable.just<Maybe<*>>(Maybe.fromSingle(Single.fromObservable(observable)))

            if (methodType == Flowable::class.java)
                return@Supplier Observable.just(observable.toFlowable(BackpressureStrategy.MISSING))

            throw RuntimeException(method.name + " needs to return one of the next reactive types: observable, single, maybe or flowable")
        }).blockingFirst()
    }
}
